kratos/solving_strategies/strategies/residualbased_newton_raphson_strategy.h

```diff
@@ -193,33 +185,145 @@ class ResidualBasedNewtonRaphsonStrategy
 
-        // Set flags to default values
-        SetMaxIterationNumber(MaxIterations);
-        mCalculateReactionsFlag = CalculateReactions;
+        // Tells to the builder and solver if the reactions have to be Calculated or not
+        GetBuilderAndSolver()->SetCalculateReactionsFlag(mCalculateReactionsFlag);
 
-        mReformDofSetAtEachStep = ReformDofSetAtEachStep;
+        // Tells to the Builder And Solver if the system matrix and vectors need to
+        //be reshaped at each step or not
+        GetBuilderAndSolver()->SetReshapeMatrixFlag(mReformDofSetAtEachStep);
 
-        // Saving the convergence criteria to be used
-        mpConvergenceCriteria = pNewConvergenceCriteria;
+        // Set EchoLevel to the default value (only time is displayed)
+        SetEchoLevel(1);
 
-        // Saving the scheme
-        mpScheme = pScheme;
+        // By default the matrices are rebuilt at each iteration
+        this->SetRebuildLevel(2);
 
-        // Saving the linear solver
-        mpLinearSolver = pNewLinearSolver;
+        mpA = TSparseSpace::CreateEmptyMatrixPointer();
+        mpDx = TSparseSpace::CreateEmptyVectorPointer();
+        mpb = TSparseSpace::CreateEmptyVectorPointer();
+
+        KRATOS_CATCH("")
+    }
+
+
+    /**
+     * Constructor with Parameters
+     * @param rModelPart The model part of the problem
+     * @param pScheme The integration scheme
+     * @param pNewLinearSolver The linear solver employed
+     * @param pNewConvergenceCriteria The convergence criteria employed
+     * @param Parameters Settings used in the strategy
+     * @param CalculateReactions The flag for the reaction calculation
+     * @param ReformDofSetAtEachStep The flag that allows to compute the modification of the DOF
+     * @param MoveMeshFlag The flag that allows to move the mesh
+     */
+    ResidualBasedNewtonRaphsonStrategy(
+        ModelPart& rModelPart,
+        typename TSchemeType::Pointer pScheme,
+        typename TLinearSolver::Pointer pNewLinearSolver,
+        typename TConvergenceCriteriaType::Pointer pNewConvergenceCriteria,
+        Parameters Settings,
+        bool CalculateReactions = false,
+        bool ReformDofSetAtEachStep = false,
+        bool MoveMeshFlag = false)
+        : SolvingStrategy<TSparseSpace, TDenseSpace, TLinearSolver>(rModelPart, MoveMeshFlag),
+          mpLinearSolver(pNewLinearSolver),
+          mpScheme(pScheme),
+          mpConvergenceCriteria(pNewConvergenceCriteria),
+          mReformDofSetAtEachStep(ReformDofSetAtEachStep),
+          mCalculateReactionsFlag(CalculateReactions),
+          mSolutionStepIsInitialized(false),
+          mInitializeWasPerformed(false),
+          mKeepSystemConstantDuringIterations(false),
+          mSettings(Settings)
+    {
+        KRATOS_TRY;
+
+        Parameters default_settings = GetDefaultSettings();
+        mSettings.AddMissingParameters(default_settings);
+        AssignSettings();
```

Comment:
you miss validation

Reply:
There is a problem with this in cpp. When you use here GetDefaultSettings() it always calls the method from the base class even if the object is a LinearSearchStrategy (derived class) because the constructor of the derived class has not been called yet. Problem is then you are validating against the base class defaults and any setting you pass that is only in the derived class will cause an error.

Reply:
make the GetDefault method static. this way you can do

BaseType::GetDefault ...

Reply:
Besides, you can even use a Lambda function to do the check while construction but keeping the member variables as `const` variables.

Reply:
@RiccardoRossi The problem here is that you need to call the GetDefaultSettings of the derived class from the constructor class of the base class, which is not possible at construction time. @rubenzorrilla I am not really familiar with Lambda functions, is there anywhere where you do something similar?

Reply:
Maybe a solution similar to the one in the `CalculateEmbeddedNodalVariableFromSkinProcess` can do the job.

Reply:
@rubenzorrilla Maybe I am missing the point but if I use that approach, I can send to the base class the parameters validated in the derived class (which also includes the values from the base class) but still I cannot validate the parameters inside the base class constructor as it does not know the derived default settings. This would only change the moment when the settings are validated (before or after the base constructor is called)

Reply:
> @rubenzorrilla Maybe I am missing the point but if I use that approach, I can send to the base class the parameters validated in the derived class (which also includes the values from the base class) but still I cannot validate the parameters inside the base class constructor as it does not know the derived default settings. This would only change the moment when the settings are validated (before or after the base constructor is called)

I think you're right. I thought that it would solve your issue. Sorry for the noise.

Reply:
I think we are at a dead end unless someone has a better solution (@pooyan-dadvand @loumalouomega  @philbucher ?) but I think this is a good example that shows why AddMissingParameters was needed..

Reply:
Can we added it?

Reply:
whay you mean @loumalouomega ?

Reply:
OK, I see is already there